File: packages/sdk/ts/src/http.ts
/*
 * Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
 * Caracal, a product of Garudex Labs
 *
 * HTTP server middleware that binds Caracal context after a verifier boundary.
 */

import type { Caracal, BindOptions } from './client.js'

export interface IncomingLike {
  headers: Record<string, string | string[] | undefined>
}

export type FastifyRequestLike = IncomingLike

export interface ServerResponseLike {
  statusCode: number
  setHeader(name: string, value: string): void
  end(body?: string): void
}

export interface FastifyReplyLike {
  code(statusCode: number): FastifyReplyLike
  header(name: string, value: string): FastifyReplyLike
  send(payload?: string): unknown
}

export type ConnectMiddleware = (req: IncomingLike, res: ServerResponseLike, next: (err?: unknown) => void) => void

const REJECT_BODY = '{"error":"unauthorized","error_description":"invalid or missing authorization"}'

export function caracalContextMiddleware(caracal: Caracal, opts: BindOptions = {}): ConnectMiddleware {
  return caracal.contextMiddleware(opts)
}

export function caracalFastifyHook(caracal: Caracal, opts: BindOptions = {}) {
  return (req: FastifyRequestLike, reply: FastifyReplyLike, done: (err?: unknown) => void): void => {
    let entered = false
    caracal
      .bindFromHeaders(
        req.headers,
        async () => {
          entered = true
          done()
        },
        opts,
      )
      .catch((err) => {
        if (entered) {
          done(err)
          return
        }
        reply.code(401).header('content-type', 'application/json').send(REJECT_BODY)
      })
  }
}

File: packages/sdk/go/http.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Context net/http middleware that binds CaracalContext after a verifier boundary.

package sdk

import (
	"log/slog"
	"net/http"
	"os"
)

// ContextMiddleware returns an http.Handler middleware that binds a CaracalContext
// from inbound envelope headers after token verification.
func (c *Caracal) ContextMiddleware(next http.Handler, opts ...CallOptions) http.Handler {
	if (len(opts) == 0 || opts[0].Verify == nil) && os.Getenv("CARACAL_ENV") == "production" {
		slog.Warn("caracal: inbound context is being bound without a verify hook in production; the envelope is propagation-only - set CallOptions.Verify or keep this boundary behind a verifier such as the Gateway")
	}
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ctx, err := c.BindFromRequest(r.Context(), r, opts...)
		if err != nil {
			http.Error(w, "invalid or missing authorization", http.StatusUnauthorized)
			return
		}
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}
